jstz_api/src/encoding/text_encoder.rs

```diff
@@ -0,0 +1,279 @@
+
+        //  3. Let encoder be an instance of the UTF-8 encoder
+        //  NB: Noop, we only support UTF-8 encoder
+        //  4. Let unused be the I/O queue of scalar values << end-of-queue >>.
+        //  NB: The handler algorithm invoked below requires this argument, but it is not used by the UTF-8 encoder.
+        //  Therefore Noop
+        //  5. Convert src to an I/O queue of scalar values.
+        //  NB: here we just convert a u16 slice to a string slice using our custom decoder
+        //      it will stop after the first unpaired surrogate
+        let source = Self::custom_decode_utf16(source);
+        let input = source.as_str();
+        //  6. While true: ...
+        //  NB: this is performed by the rust `encoding_rs` crate
+        let (result, _, has_error) = encoding_rs::UTF_8.encode(input);
+        assert!(!has_error);
+        destination.append(&mut result.to_vec());
+        TextEncoderEncodeIntoResult { read, written }
+    }
+}
+#[derive(Default, Clone, Trace, Finalize)]
+pub struct TextEncoderClass;
+impl TextEncoderClass {
+    fn encoding(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextEncoder,
+            "encoding",
+            get:((this, _context) => Ok(JsString::from(this.encoding()).into()))
+        )
+    }
+
+    fn encode(_: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        let arg = args.get_or_undefined(0).as_string().map(|x| x.as_slice());
+        let result = TextEncoder::encode(arg);
+        let byte_block = result.to_vec();
+        let array_buffer: JsArrayBuffer =
+            JsArrayBuffer::from_byte_block(byte_block, context)?;
+        let result_js = JsUint8Array::from_array_buffer(array_buffer, context)?;
+        Ok(result_js.into())
+    }
+
+    fn encode_into(
+        _: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        let source = args.get_or_undefined(0);
+        let source = source
+            .as_string()
+            .map(|x| x.as_slice())
+            .ok_or::<boa_engine::JsError>(
+                JsNativeError::typ()
+                    .with_message("Failed to interpret argument as a string")
+                    .into(),
+            )?;
+        let destination = args.get_or_undefined(1);
+        // 2nd argument is a reference to a Uint8Array to be overwritten
+        let init: JsUint8Array = destination.try_js_into(context)?;
+        // initialize a boxed "destination buffer" with the transpiled content of the Js object
+        let destination: &mut Vec<u8> = &mut Vec::new();
+        // call the rust function
+        let result: TextEncoderEncodeIntoResult =
+            TextEncoder::encode_into(source, destination);
+        // write back the destination JsValue
+        for (i, value) in destination.iter().enumerate() {
+            init.fill(*value, Some(i), Some(i + 1), context)?;
+        }
```

Comment:
I think it is possible to pass the underlying reference from the `init` to `encode_into`. I believe I explained this in 1 of the comments in one of the closed PRs for text encoder

Reply:
comment : https://github.com/trilitech/jstz/pull/83#discussion_r1376572314

Reply:
done (in the last commit).
But I am too bad at rust to do it cleanly